File: C++/minimum-sum-after-divisible-sum-deletions.cpp
// Time:  O(n + k)
// Space: O(k)

// dp, prefix sum
class Solution {
public:
    long long minArraySum(vector<int>& nums, int k) {
        vector<int64_t> dp(k, numeric_limits<int64_t>::max());
        dp[0] = 0;
        int64_t result = 0;
        for (const auto& x : nums) {
            result += x;
            dp[result % k] = result = min(result, dp[result % k]);
        }
        return result;
    }
};

File: Python/minimum-sum-after-divisible-sum-deletions.py
# Time:  O(n + k)
# Space: O(k)

# dp, prefix sum
class Solution(object):
    def minArraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        dp = [float("inf")]*k
        dp[0] = result = 0
        for x in nums:
            result += x
            dp[result%k] = result = min(result, dp[result%k])
        return result
